pkg/apis/resource/types.go

```diff
@@ -17,17 +17,288 @@ limitations under the License.
+//
+// Whenever a driver needs to update a pool, it increments the pool.Spec.Pool.Generation number
+// and updates all ResourceSlices with that new number and new resource definitions. A consumer
+// must only use ResourceSlices with the highest generation number and ignore all others.
+//
+// When allocating all resources in a pool matching certain criteria or when
+// looking for the best solution among several different alternatives, a
+// consumer should check the number of ResourceSlices in a pool (included in
+// each ResourceSlice) to determine whether its view of a pool is complete and
+// if not, should wait until the driver has completed updating the pool.
+//
+// For resources that are not local to a node, the node name is not set. Instead,
+// the driver may use a node selector to specify where the devices are available.
+//
+// This is an alpha type and requires enabling the DynamicResourceAllocation
+// feature gate.
+type ResourceSlice struct {
+	metav1.TypeMeta
+	// Standard object metadata
+	// +optional
+	metav1.ObjectMeta
+
+	// Contains the information published by the driver.
+	//
+	// Changing the spec automatically increments the metadata.generation number.
+	Spec ResourceSliceSpec
+}
+
+const (
+	// ResourceSliceSelectorNodeName can be used in a [metav1.ListOptions]
+	// field selector to filter based on [ResourceSliceSpec.NodeName].
+	ResourceSliceSelectorNodeName = "spec.nodeName"
+	// ResourceSliceSelectorDriver can be used in a [metav1.ListOptions]
+	// field selector to filter based on [ResourceSliceSpec.Driver].
+	ResourceSliceSelectorDriver = "spec.driver"
+)
+
+// ResourceSliceSpec contains the information published by the driver in one ResourceSlice.
+type ResourceSliceSpec struct {
+	// Driver identifies the DRA driver providing the capacity information.
+	// A field selector can be used to list only ResourceSlice
+	// objects with a certain driver name.
+	//
+	// Must be a DNS subdomain and should end with a DNS domain owned by the
+	// vendor of the driver. This field is immutable.
+	//
+	// +required
+	Driver string
+
+	// Pool describes the pool that this ResourceSlice belongs to.
+	//
+	// +required
+	Pool ResourcePool
+
+	// NodeName identifies the node which provides the resources in this pool.
+	// A field selector can be used to list only ResourceSlice
+	// objects belonging to a certain node.
+	//
+	// This field can be used to limit access from nodes to ResourceSlices with
+	// the same node name. It also indicates to autoscalers that adding
+	// new nodes of the same type as some old node might also make new
+	// resources available.
+	//
+	// Exactly one of NodeName, NodeSelector and AllNodes must be set.
+	// This field is immutable.
+	//
+	// +optional
+	// +oneOf=NodeSelection
+	NodeName string
+
+	// NodeSelector defines which nodes have access to the resources in the pool,
+	// when that pool is not limited to a single node.
+	//
+	// Must use exactly one term.
```

Comment:
From https://github.com/kubernetes/kubernetes/pull/125488#issuecomment-2230877076:

> I cannot reproduce locally. Might be timing issues. I keep looking...

Once I started testing with a multi-node cluster, I could reproduce. The good news: E2E test coverage of structure parameters is good enough to catch an incomplete implementation. The bad news: I forgot to initialize the node selector of an allocated claim in `allocator.go`, so pods did not get scheduled to the right node in a multi-node cluster...

While implementing this, I found that in order to "join" two NodeSelectors from different slices into one for the claim,  it's easier when imposing the limitation that the individual slice NodeSelectors must have exactly one term. Then I can simply add together the requirements in that term and get one NodeSelector with again only one term. Otherwise I need to build the cross-product of all terms, which becomes very unwieldy - not necessarily impossible, but probably not desirable.

Example: `"country" in "us"` in slice1 and `"region" in "west"` in slice2 becomes `["country" in "us", "region" in "west"]` in the claim node selector.

But when one slice uses multiple terms, they are ORed, leading to: `"country" in "us"` in slice1, `"region" in "west" OR "city" in "New York"` => `["country" in "us", "region" in "west"] OR ["country" in "us", "city" in "New York"]`

I think this is a reasonable limitation. If it becomes a problem, we can still relax later.

@thockin: this is checked in the updated `validation.go`, see [diff](https://github.com/kubernetes/kubernetes/compare/40a45e61de52b44921b9c8d40876d61909dc2434..83c8688f9da030edf5cc10135def0ec70c2056f2) of the latest force-push.

@johnbelamaric, @bart0sh: `allocator.go` and `allocator_test.go` were updated accordingly.